File: VTabView.m
#import "VTabView.h"
#import "VTabController.h"
#import "VTablature.h"
#import "VNote.h"
#import "TLSelectionManager.h"

#define STRING_SPACE 14.0
#define LINE_SPACE 24.0
#define LEFT_MARGIN 16.0
#define RIGHT_MARGIN 16.0
#define TOP_MARGIN 16.0
#define LINE_WIDTH 2
#define CHORD_SPACE 24.0

@interface VTabView (Private)

// drawing submethods
- (void)drawStringsWithGraphicsContext:(NSGraphicsContext *)theContext;
- (void)drawTabWithGraphicsContext:(NSGraphicsContext *)theContext;
- (void)drawFocusRectAtPoint:(NSPoint)origin
                      ofSize:(NSSize)size
                     inColor:(NSColor *)strokeColor;

// drawing helper functions
- (NSUInteger)lineHeight;
- (NSUInteger)chordsPerLine;

// internal information
- (NSInteger)chordIndexAtPoint:(NSPoint)thePoint;
- (NSInteger)stringAtPoint:(NSPoint)thePoint;

@end

@implementation VTabView

@synthesize tablature;
@synthesize selectionManager;
@synthesize lastFocusChordIndex;
@synthesize focusChordIndex;
@synthesize focusNoteString;
@synthesize mouseDownEvent;

#pragma mark -
#pragma mark Drawing Functions

- (CGFloat)drawOneLineOfStringsAtHeight:(CGFloat)stringHeight
                           withSpaceFor:(NSUInteger)lineLength
{
    NSPoint startPoint;
    NSPoint endPoint;
    NSUInteger stringNum;
    [NSBezierPath setDefaultLineWidth:LINE_WIDTH];
    [[NSColor lightGrayColor] setStroke];
    for (stringNum = 0;
         stringNum < [tablature numStrings] && stringHeight < [self bounds].size.height;
         stringNum++) {
        // draw the strings
        startPoint = NSMakePoint(LEFT_MARGIN,
                                 stringHeight);
        endPoint = NSMakePoint(LEFT_MARGIN + (lineLength * CHORD_SPACE), 
                               stringHeight);
        [NSBezierPath strokeLineFromPoint:startPoint toPoint:endPoint];
        stringHeight += STRING_SPACE;
    }
    return stringHeight;
}

- (void)drawFocusRectForChordAtPoint:(NSPoint)origin
                             inColor:(NSColor *)strokeColor
{
    NSSize size;
    CGFloat rectRadius;
    if ([tabController isInSoloMode]) {
        size = NSMakeSize(1.0, [self lineHeight]);
        origin = NSMakePoint(origin.x + CHORD_SPACE - 1.0, origin.y);
        rectRadius = 1.0;
    }
    else {
        size = NSMakeSize(CHORD_SPACE, [self lineHeight]);
        rectRadius = 3.0;
    }
    NSRect focusRect = {origin, size};
    NSBezierPath *focusPath = [NSBezierPath bezierPathWithRoundedRect:focusRect
                                                              xRadius:rectRadius
                                                              yRadius:rectRadius];
    [[strokeColor colorWithAlphaComponent:0.5] setStroke];
    [focusPath stroke];
}

- (void)drawChord:(VChord *)chord
     withCornerAt:(NSPoint)topLeft
      normalStyle:(NSDictionary *)tabAttrs
     focusedStyle:(NSDictionary *)focusNoteAttrs
{
    for (VNote *note in chord) {
        bool focused = (note == [self focusNote]);
        NSDictionary *attrsToUse = focused ? focusNoteAttrs : tabAttrs;
        NSString *text = [note hasFret] ? [note stringValue] : @"•";
        if ([note hasFret] || focused) {
            [text drawAtPoint:NSMakePoint(topLeft.x + CHORD_SPACE/3, topLeft.y)
               withAttributes:attrsToUse];
        }
        topLeft.y += STRING_SPACE;
    }
}

- (void)drawSelectionForChordRange:(NSRange)chordRange
                     withTopLeftAt:(NSPoint)topLeft
                        usingColor:(NSColor *)selectionColor
{
    NSRect selectRect;
    NSIndexSet *thisRangeSelection = [[selectionManager selectedIndexes]
                                      indexesInRange:chordRange
                                             options:0
                                         passingTest:^(NSUInteger idx, BOOL *stop) { return YES; }];
    NSRange thisRowRange = NSMakeRange([thisRangeSelection firstIndex] - chordRange.location,
                                       [thisRangeSelection lastIndex] - [thisRangeSelection firstIndex] + 1);

    selectRect.origin = NSMakePoint(topLeft.x + thisRowRange.location * CHORD_SPACE,
                                    topLeft.y);
    selectRect.size = NSMakeSize(thisRowRange.length * CHORD_SPACE, [self lineHeight]);
    NSBezierPath *selectionPath = [NSBezierPath bezierPathWithRoundedRect:selectRect
                                                                  xRadius:3.0
                                                                  yRadius:3.0];
    [[selectionColor colorWithAlphaComponent:0.3] setFill];
    [selectionPath fill];
}

- (void)drawOneLineOfTabAtHeight:(CGFloat)tabHeight
                 fromChordNumber:(NSUInteger)firstChord
                  numberOfChords:(NSUInteger)numChords
{
    NSPoint currentCoords = NSMakePoint(LEFT_MARGIN, tabHeight);
    NSUInteger chordNum = firstChord;
    NSDictionary *tabAttrs = [NSDictionary dictionary];
    NSMutableDictionary *focusNoteAttrs = [NSMutableDictionary dictionaryWithDictionary:tabAttrs];
//    NSFontManager *fontManager = [NSFontManager sharedFontManager];
//    NSFont *boldUserFont = [fontManager convertFont:[NSFont userFontOfSize:12.0]
//                                        toHaveTrait:NSBoldFontMask];
//    [focusNoteAttrs setValue:boldUserFont
//                      forKey:NSFontAttributeName];
    [focusNoteAttrs setValue:[NSColor redColor]
                      forKey:NSForegroundColorAttributeName];
    [focusNoteAttrs setValue:[NSNumber numberWithFloat:-6.0]
                      forKey:NSStrokeWidthAttributeName];
//    NSForegroundColorAttributeName : [NSColor redColor]
//    NSStrokeWidthAttributeName : -3.0
    NSColor *selectionColor = [NSColor blueColor];
    [selectionColor set];
    [self drawSelectionForChordRange:NSMakeRange(firstChord, numChords)
                       withTopLeftAt:currentCoords
                          usingColor:selectionColor];
    for (chordNum = firstChord; chordNum < firstChord + numChords; ++chordNum) {
        VChord *chord = [tablature objectInChordsAtIndex:chordNum];
        [self drawChord:chord
           withCornerAt:currentCoords
            normalStyle:tabAttrs
           focusedStyle:focusNoteAttrs];
        currentCoords.y = tabHeight;
        if (chord == [self focusChord]) {
            [self drawFocusRectForChordAtPoint:currentCoords
                                       inColor:selectionColor];
        }
        currentCoords.x += CHORD_SPACE;
    }
}

- (void)drawTabWithGraphicsContext:(NSGraphicsContext *)theContext
{
    CGFloat stringHeight = TOP_MARGIN;
    CGFloat tabHeight = TOP_MARGIN;
    NSUInteger chordsAccommodated = 0;
    NSUInteger lineLength = 0;
    const NSUInteger chordsPerLine = (NSUInteger)(([self bounds].size.width - LEFT_MARGIN - RIGHT_MARGIN) / 
                                                  CHORD_SPACE);
    
    do {
        // draw one line's worth of tab strings
        if (chordsAccommodated + chordsPerLine > [tablature countOfChords]) {
            // set length to accommodate remaining chords
            lineLength = [tablature countOfChords] - chordsAccommodated;
        }
        else {
            // set length to max based on view width
            lineLength = chordsPerLine;
        }
        tabHeight = stringHeight - [[NSFont userFontOfSize:12.0] xHeight];
        stringHeight = [self drawOneLineOfStringsAtHeight:stringHeight
                                             withSpaceFor:lineLength] + LINE_SPACE;
        [self drawOneLineOfTabAtHeight:tabHeight
                       fromChordNumber:chordsAccommodated
                        numberOfChords:lineLength];
        chordsAccommodated += lineLength;
    } while (chordsAccommodated < [tablature countOfChords] && 
             stringHeight + [self lineHeight] <= [self bounds].size.height);
}

- (void)drawRect:(NSRect)dirtyRect
{
    NSGraphicsContext *startGraphicsContext = [NSGraphicsContext currentContext];
    [startGraphicsContext saveGraphicsState];
    [[NSColor whiteColor] setFill];
    NSRectFill(dirtyRect);
    [self drawTabWithGraphicsContext:startGraphicsContext];
    [startGraphicsContext restoreGraphicsState];
}

- (BOOL)isFlipped
{
    return YES;
}

- (BOOL)acceptsFirstResponder
{
    return YES;
}

- (void)awakeFromNib
{
    [self setSelectionManager:[TLSelectionManager new]];
    [selectionManager setDelegate:self];
    [tabController setNextResponder:[self nextResponder]];
    [self setNextResponder:tabController];
    if ([tablature countOfChords] >= 1) {
        [selectionManager selectIndexes:[NSIndexSet indexSetWithIndex:0]
                   byExtendingSelection:NO];
        [self setFocusChordIndex:0];
        [self setFocusNoteString:0];
    }
    [self setNeedsDisplay:YES];
    return;
}

#pragma mark -
#pragma mark Selection handling

- (BOOL)hasSelection
{
    return ([[selectionManager selectedIndexes] count] > 0);
}

- (NSIndexSet *)selectedIndexes
{
    return [selectionManager selectedIndexes];
}

- (NSArray *)selectedChords
{
    return [tablature chordsAtIndexes:[self selectedIndexes]];
}

- (void)selectIndexes:(NSIndexSet *)indexes
{
    [selectionManager selectIndexes:indexes
               byExtendingSelection:NO];
}

- (void)clearSelection
{
    [selectionManager selectIndexes:[NSIndexSet indexSet]
               byExtendingSelection:NO];
}

- (VChord *)focusChord
{
    return [tablature objectInChordsAtIndex:focusChordIndex];
}

- (VNote *)focusNote
{
    return [[self focusChord] objectInNotesAtIndex:focusNoteString];
}

- (NSUInteger)chordsPerLine
{
    return (NSUInteger)(([self bounds].size.width - LEFT_MARGIN - RIGHT_MARGIN) / CHORD_SPACE);
}

- (NSUInteger)lineHeight
{
    return (NSUInteger)(([tablature numStrings] - 0.5) * (STRING_SPACE + LINE_WIDTH));
}

- (VChord *)chordAtPoint:(NSPoint)thePoint
{
    NSInteger chordIndex = [self chordIndexAtPoint:thePoint];
    if (([tablature countOfChords] > chordIndex) &&
        (chordIndex != NO_HIT)) {
        return [tablature objectInChordsAtIndex:chordIndex];
    } else {
        return nil;    
    }
}

- (NSInteger)chordIndexAtPoint:(NSPoint)thePoint
{
    CGFloat x = thePoint.x - LEFT_MARGIN;
    CGFloat y = thePoint.y - TOP_MARGIN;
    // number of full lines of tab above click location
    NSUInteger skipLines = (int)(y / ([self lineHeight] + LINE_SPACE));
    if (fmod(y, skipLines) > [self lineHeight]) {
        return NO_HIT;
    }
    else {
        NSUInteger skipChords = (int)(x / CHORD_SPACE);
        NSUInteger chordIndex = skipLines * [self chordsPerLine] + skipChords;
        NSLog(@"Chord index: %lu", chordIndex);
        return chordIndex;
    }
}

- (VNote *)noteAtPoint:(NSPoint)thePoint
{
    VChord *theChord;
    if ((theChord = [self chordAtPoint:thePoint])) {
        return [theChord objectInNotesAtIndex:[self stringAtPoint:thePoint]];
    }
    else {
        return nil;
    }
}

- (NSInteger)stringAtPoint:(NSPoint)thePoint
{
    CGFloat y = thePoint.y - TOP_MARGIN;
    // find y-position of click relative to the top of its tab line
    y -= (int)(y / ([self lineHeight] + LINE_SPACE)) * ([self lineHeight] + LINE_SPACE);
    NSUInteger whichString = MIN((int)(y / STRING_SPACE), [tablature numStrings] - 1);
    NSLog(@"Clicked string: %lu, y = %f", whichString, y);
    return whichString;
}

#pragma mark -
#pragma mark TLSelectionList delegate method

- (NSUInteger)selectionManager:(TLSelectionManager *)manager
               indexUnderPoint:(NSPoint)windowPoint
                      userInfo:(void *)userInfo
{
    return [self chordIndexAtPoint:[self convertPoint:windowPoint
                                             fromView:nil]];
}

//- (BOOL)selectionManagerShouldInitiateDragLater:(TLSelectionManager*)manager
//									  dragEvent:(NSEvent*)dragEvent
//								  originalEvent:(NSEvent*)mouseDownEvent
//									   userInfo:(void*)userInfo
//{
//    return YES;
//}

- (NSIndexSet *)selectionManager:(TLSelectionManager *)manager
                    indexesInBox:(NSRect)windowRect
                        userInfo:(void *)userInfo
{
    NSRect viewRect = [self convertRect:windowRect
                               fromView:nil];
    NSPoint topLeft = viewRect.origin;
    NSPoint topRight = NSMakePoint(topLeft.x + viewRect.size.width, viewRect.origin.y);
    NSPoint bottomLeft = NSMakePoint(viewRect.origin.x, topLeft.y + viewRect.size.height);
    NSPoint bottomRight = NSMakePoint(topLeft.x + viewRect.size.width,
                                      topLeft.y + viewRect.size.height);
    
    NSPoint mouseDownPoint = [self convertPoint:[[self mouseDownEvent] locationInWindow]
                                       fromView:nil];
    NSInteger startIndex, endIndex;
    if (NSEqualPoints(mouseDownPoint, topLeft) || NSEqualPoints(mouseDownPoint, bottomRight)) {
        startIndex = MAX([self chordIndexAtPoint:topLeft], 0);
        endIndex = MIN([self chordIndexAtPoint:bottomRight], [tablature countOfChords] - 1);
    }
    else {
        startIndex = MAX([self chordIndexAtPoint:topRight], 0);
        endIndex = MIN([self chordIndexAtPoint:bottomLeft], [tablature countOfChords] - 1);
    }
    
    return [NSIndexSet indexSetWithIndexesInRange:NSMakeRange(MIN(startIndex, endIndex),
                                                              abs(endIndex - startIndex) + 1)];
}

- (BOOL)shouldResetExistingSelection {
	return YES;
}

#pragma mark -
#pragma mark Input Handling

- (void)handleBoundKey:(NSString *)keyString
{
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Warc-performSelector-leaks"
    NSArray *actionParts = [[tabController keyBindings] objectForKey:keyString];
    NSString *selectorString = [actionParts objectAtIndex:0];
    SEL editSelector = NSSelectorFromString(selectorString);
    switch ([actionParts count])
    {
        case 1:
            [tabController performSelector:editSelector];
            break;
        case 2:
            if ([selectorString isEqualToString:@"addOpenString:"]) {
                [tabController addOpenString:[actionParts objectAtIndex:1]
                               reverseString:YES];
            }
            else {
                [tabController performSelector:editSelector
                                    withObject:[actionParts objectAtIndex:1]];
            }
            break;
        case 3:
            if ([selectorString isEqualToString:@"addNoteOnString:onFret:"]) {
                [tabController addNoteOnString:[actionParts objectAtIndex:1]
                                        onFret:[actionParts objectAtIndex:2]
                                 reverseString:YES];
            }
            else {
                [tabController performSelector:editSelector
                                    withObject:[actionParts objectAtIndex:1]
                                    withObject:[actionParts objectAtIndex:2]];
            }
    }
#pragma clang diagnostic pop
}

- (void)keyDown:(NSEvent *)theEvent
{
    if ([[tabController keyBindings] objectForKey:[theEvent characters]]) {
        [self handleBoundKey:[theEvent characters]];
    }
    else {
        [self interpretKeyEvents:[NSArray arrayWithObject:theEvent]];
    }
    [self setNeedsDisplay:YES];
}

- (void)reFocusAtPoint:(NSPoint)thePoint
{
    focusChordIndex = [self chordIndexAtPoint:thePoint];
    if (focusChordIndex >= [tablature countOfChords]) {
        focusChordIndex = [tablature countOfChords] - 1;
    }
    
    focusNoteString = [self stringAtPoint:thePoint];
    [self setNeedsDisplay:YES];
}

- (void)mouseDown:(NSEvent*)theEvent
{
    [self setLastFocusChordIndex:[self focusChordIndex]];
    [self setMouseDownEvent:theEvent];
    [selectionManager mouseDown:theEvent userInfo:NULL];
    [self reFocusAtPoint:[self convertPoint:[theEvent locationInWindow]
                                   fromView:nil]];
}

- (void)mouseDragged:(NSEvent*)theEvent
{
    [selectionManager mouseDragged:theEvent userInfo:NULL];
    [self reFocusAtPoint:[self convertPoint:[theEvent locationInWindow]
                                   fromView:nil]];
}

- (void)mouseUp:(NSEvent*)theEvent
{
    [selectionManager mouseUp:theEvent];
    if ([[selectionManager selectedIndexes] count] == 1 &&
        [[selectionManager selectedIndexes] firstIndex] != [self lastFocusChordIndex]) {
        [self clearSelection];
    }
    [self reFocusAtPoint:[self convertPoint:[theEvent locationInWindow]
                                   fromView:nil]];
}

- (void)selectionManagerDidChangeSelection:(TLSelectionManager*)manager
{
    [self setNeedsDisplay:YES];
}

- (void)focusNextChord
{
    focusChordIndex++;
    [self clearSelection];
}

- (void)focusPrevChord
{
    focusChordIndex--;
    [self clearSelection];
}

- (void)focusUpString
{
    focusNoteString--;
}

- (void)focusDownString
{
    focusNoteString++;
}

@end

File: VTabView.swift
//
//  VTabView.swift
//  tablution
//
//  Created by Ian McCowan on 11/1/15.
//
//

import Foundation

let LINE_SPACE: CGFloat = 24.0
let LEFT_MARGIN: CGFloat = 16.0
let TAB_HEADER_WIDTH: CGFloat = 16.0
let RIGHT_MARGIN: CGFloat = 16.0
let TOP_MARGIN: CGFloat = 16.0
let LINE_WIDTH: CGFloat = 2.0
let CHORD_SPACE: CGFloat = 24.0

@objc public class VTabView: NSView {
    @IBOutlet weak var tabController: VTabController?
    var selectionManager: TLSelectionManager?
    var lastFocusChordIndex: Int = 0
    var currFocusChordIndex: Int = 0
    var focusNoteString: Int = 0
    var mouseDownEvent: NSEvent?
    var tablature: VTablature?
    var tabFont: NSFont = NSFont.systemFontOfSize(12.0)
    
    public override var acceptsFirstResponder: Bool {
        get { return true }
    }
    public override var flipped: Bool {
        get { return true }
    }
    
    public override init(frame frameRect: NSRect) {
        super.init(frame: frameRect)
        guard let selectionManager = TLSelectionManager() as TLSelectionManager? else {
            fatalError("Couldn't initialize selection manager")
        }
        selectionManager.delegate = self
        selectionManager.selectIndexes(NSIndexSet(), byExtendingSelection: false)
        self.selectionManager = selectionManager
    }

    required public init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override public func awakeFromNib() {
        tabController!.nextResponder = self.window
        self.nextResponder = tabController
        currFocusChordIndex = 0
        focusNoteString = 0
        needsDisplay = true
        
        // set tab font from defaults
        let defaults: NSUserDefaults = NSUserDefaults.standardUserDefaults()
        let tabFontDefault = defaults.valueForKey("tabFont")
        guard tabFontDefault != nil else {
            NSLog("Default tab font not set yet. Aborting drawing.")
            return
        }
        guard let fontData = tabFontDefault as? NSData else {
            fatalError("tabFont user default must be an NSData-encoded NSFont")
        }
        if let tabFontFromDefaults = NSKeyedUnarchiver.unarchiveObjectWithData(fontData) as? NSFont {
            self.tabFont = tabFontFromDefaults
        }
        else {
            fatalError("tabFont user default must be an NSData-encoded NSFont")
        }
    }
    
    private func drawOneLineOfStringsAtHeight(stringHeight: CGFloat, withSpaceFor lineLength: Int) -> CGFloat {
        var startPoint: NSPoint
        var endPoint: NSPoint
        var stringNum: Int
        var newStringHeight: CGFloat = stringHeight
        NSBezierPath.setDefaultLineWidth(LINE_WIDTH)
        NSColor.lightGrayColor().setStroke()
        for stringNum = 0; stringNum < tablature!.numStrings && stringHeight < self.bounds.size.height; stringNum++ {
            startPoint = NSMakePoint(LEFT_MARGIN + TAB_HEADER_WIDTH, newStringHeight)
            endPoint = NSMakePoint(LEFT_MARGIN + TAB_HEADER_WIDTH + (CGFloat(lineLength) * CHORD_SPACE), newStringHeight)
            NSBezierPath.strokeLineFromPoint(startPoint, toPoint: endPoint)
            newStringHeight += tabFont.pointSize
        }
        return stringHeight
    }

    func drawFocusRectForChordAtPoint(origin: NSPoint, inColor strokeColor: NSColor) {
        var size: NSSize
        var rectRadius: CGFloat
        var newOrigin: NSPoint = origin
        if tabController!.isInSoloMode() {
            size = NSMakeSize(2.0, self.lineHeight())
            newOrigin = NSMakePoint(origin.x - 1.0, origin.y)
            rectRadius = 1.0
        }
        else {
            size = NSMakeSize(CHORD_SPACE, self.lineHeight())
            rectRadius = 3.0
        }
        let focusRect: NSRect = NSRect(origin: newOrigin, size: size)
        let focusPath: NSBezierPath = NSBezierPath(roundedRect: focusRect, xRadius: rectRadius, yRadius: rectRadius)
        strokeColor.colorWithAlphaComponent(0.5).setStroke()
        focusPath.stroke()
    }
    
    func drawChord(chord: VChord, withCornerAt topLeft: NSPoint, normalStyle tabAttrs: [String : AnyObject], focusedStyle focusNoteAttrs: [String : AnyObject]) {
        // This could maybe be done by drawing the chord as a single string with linebreaks, rather than note by note.
        // But it would require some shenanigans with NSMutableAttributedString to get the focused note to work.
        var topLeftForDrawing = topLeft
        for note in chord.notes {
            let focused: Bool = (note === self.focusNote())
            let attrsToUse: [String : AnyObject] = focused ? focusNoteAttrs : tabAttrs
            let text: String = note.hasFret() ? note.stringValue() : "•"
            if note.hasFret() || focused {
                text.drawAtPoint(NSMakePoint(topLeftForDrawing.x + CHORD_SPACE / 3, topLeftForDrawing.y), withAttributes: attrsToUse)
            }
            topLeftForDrawing.y += tabFont.pointSize
        }
    }
    
    func drawSelectionForChordRange(chordRange: NSRange, withTopLeftAt topLeft: NSPoint, usingColor selectionColor: NSColor) {
        let thisRangeSelection: NSIndexSet = selectionManager!.selectedIndexes.indexesInRange(chordRange, options: NSEnumerationOptions(),
            passingTest: {(idx: Int, stop: UnsafeMutablePointer<ObjCBool>) in return true })
        let thisRowRange: NSRange = NSMakeRange(thisRangeSelection.firstIndex - chordRange.location,
            thisRangeSelection.lastIndex - thisRangeSelection.firstIndex + 1)
        let selectRect: NSRect = NSRect(
            origin: NSMakePoint(topLeft.x + CGFloat(thisRowRange.location) * CHORD_SPACE, topLeft.y),
            size: NSMakeSize(CGFloat(thisRowRange.length) * CHORD_SPACE, self.lineHeight())
        )
        let selectionPath: NSBezierPath = NSBezierPath(roundedRect: selectRect, xRadius: 3.0, yRadius: 3.0)
        selectionColor.colorWithAlphaComponent(0.3).setFill()
        selectionPath.fill()
    }
    
    func drawMeasureBarRightOfPoint(topLeft: NSPoint, rightBy xDistance: CGFloat = CHORD_SPACE) {
        NSBezierPath.setDefaultLineWidth(LINE_WIDTH)
        NSColor.blackColor().setStroke()
        let startPoint: NSPoint = NSMakePoint(topLeft.x + xDistance, topLeft.y)
        let endPoint: NSPoint = NSMakePoint(topLeft.x + xDistance, topLeft.y + self.lineHeight())
        NSBezierPath.strokeLineFromPoint(startPoint, toPoint: endPoint)
    }
    
    func drawTabHeaderLine(withCornerAt topLeft: NSPoint) {
        var topLeftForDrawing = topLeft
        for label in tablature!.stringLabels.reverse() {
            label.drawAtPoint(NSMakePoint(topLeftForDrawing.x, topLeftForDrawing.y), withAttributes: [
                NSFontAttributeName:self.tabFont,
                NSStrokeWidthAttributeName:NSNumber(float: -3.0)
            ])
            topLeftForDrawing.y += tabFont.pointSize
        }
        self.drawMeasureBarRightOfPoint(topLeft, rightBy: TAB_HEADER_WIDTH)
    }
    
    func drawOneLineOfTabAtHeight(tabHeight: CGFloat, fromChordNumber firstChord: Int, numberOfChords numChords: Int) {
        var currentCoords: NSPoint = NSMakePoint(LEFT_MARGIN + TAB_HEADER_WIDTH, tabHeight)
        var chordNum: Int = firstChord
        let tabAttrs: [String : AnyObject] = [NSFontAttributeName:self.tabFont]
        var focusNoteAttrs: [String : AnyObject] = tabAttrs
        focusNoteAttrs[NSForegroundColorAttributeName] = NSColor.redColor()
        focusNoteAttrs[NSStrokeWidthAttributeName] = NSNumber(float: -6.0)
        let selectionColor: NSColor = NSColor.blueColor()
        if self.hasSelection() {
            selectionColor.set()
            self.drawSelectionForChordRange(NSMakeRange(firstChord, numChords), withTopLeftAt: currentCoords, usingColor: selectionColor)
        }
        for chordNum = firstChord; chordNum < firstChord + numChords; ++chordNum {
            let chord: VChord = tablature!.objectInChordsAtIndex(chordNum) as! VChord
            self.drawChord(chord, withCornerAt: currentCoords, normalStyle: tabAttrs, focusedStyle: focusNoteAttrs)
            if tablature!.hasBarAtIndex(chordNum) {
                self.drawMeasureBarRightOfPoint(currentCoords)
            }
            currentCoords.y = tabHeight
            if chordNum == self.currFocusChordIndex {
                self.drawFocusRectForChordAtPoint(currentCoords, inColor: selectionColor)
            }
            currentCoords.x += CHORD_SPACE
        }
        if self.currFocusChordIndex == tablature!.countOfChords() {
            self.drawFocusRectForChordAtPoint(currentCoords, inColor: selectionColor)
        }
    }
    
    func drawTab() {
        var stringHeight: CGFloat = TOP_MARGIN
        var tabHeight: CGFloat = TOP_MARGIN
        var chordsAccommodated: Int = 0
        var lineLength: Int = 0
        let chordsPerLine: Int = Int(self.chordsPerLine())
        repeat {
            if chordsAccommodated + chordsPerLine > tablature!.countOfChords() {
                lineLength = tablature!.countOfChords() - chordsAccommodated
            }
            else {
                lineLength = chordsPerLine
            }
            tabHeight = stringHeight - tabFont.xHeight
            stringHeight = self.drawOneLineOfStringsAtHeight(stringHeight, withSpaceFor: lineLength) + LINE_SPACE
            self.drawTabHeaderLine(withCornerAt: NSMakePoint(LEFT_MARGIN, tabHeight))
            self.drawOneLineOfTabAtHeight(tabHeight, fromChordNumber: chordsAccommodated, numberOfChords: lineLength)
            chordsAccommodated += lineLength
        } while chordsAccommodated < tablature!.countOfChords() && stringHeight + self.lineHeight() <= self.bounds.size.height
    }
    
    public override func changeFont(senderOrNil: AnyObject?) {
        let oldFont: NSFont = tabFont
        guard let sender = senderOrNil as? NSFontManager else {
            return
        }
        if let newFont = sender.convertFont(oldFont) as NSFont? {
            self.tabFont = newFont
        }
        self.needsDisplay = true
    }
    
    public override func drawRect(dirtyRect: NSRect) {
        guard self.tablature != nil else {
            return
        }
        NSGraphicsContext.saveGraphicsState()
        NSColor.whiteColor().setFill()
        NSRectFill(dirtyRect)
        self.drawTab()
        NSGraphicsContext.restoreGraphicsState()
    }
    
    func chordsPerLine() -> CGFloat {
        return (self.bounds.size.width - LEFT_MARGIN - RIGHT_MARGIN) / CHORD_SPACE
    }
    
    func lineHeight() -> CGFloat {
        return (CGFloat(tablature!.numStrings) - 0.5) * (tabFont.pointSize + LINE_WIDTH)
    }
    
    func chordAtPoint(thePoint: NSPoint) -> VChord? {
        if let chordIndex: Int = self.chordIndexAtPoint(thePoint) {
            if (tablature!.countOfChords() > chordIndex) {
                return tablature!.objectInChordsAtIndex(chordIndex) as? VChord
            }
        }
        return nil
    }
    
    func chordIndexAtPoint(thePoint: NSPoint) -> Int? {
        let x: CGFloat = sandwich(lower: LEFT_MARGIN, num: thePoint.x, upper: self.chordsPerLine() * CHORD_SPACE) - LEFT_MARGIN
        let y: CGFloat = sandwich(lower: TOP_MARGIN, num: thePoint.y, upper: self.bounds.size.height) - TOP_MARGIN
        let skipLines: CGFloat = (y / (self.lineHeight() + LINE_SPACE))
        if (y % skipLines) > self.lineHeight() {
            return nil
        }
        else {
            let skipChords: Int = Int(x / CHORD_SPACE)
            let chordIndex: Int = Int(skipLines * self.chordsPerLine()) + skipChords
            NSLog("Chord index: %lu", chordIndex)
            return chordIndex
        }
    }
    
    func noteAtPoint(thePoint: NSPoint) -> VNote? {
        if let theChord = self.chordAtPoint(thePoint) as VChord? {
            return theChord[self.stringAtPoint(thePoint)]
        }
        else {
            return nil
        }
    }
    
    func stringAtPoint(thePoint: NSPoint) -> Int {
        var y: CGFloat = thePoint.y - TOP_MARGIN
        y -= (y / (self.lineHeight() + LINE_SPACE)) * (self.lineHeight() + LINE_SPACE)
        let whichString: Int = min(Int(y / tabFont.pointSize), tablature!.numStrings - 1)
        NSLog("Clicked string: %lu, y = %f", whichString, y)
        return whichString
    }
    
    func hasSelection() -> Bool {
        return self.selectedIndexes().count > 0
    }
    
    func selectedIndexes() -> NSIndexSet {
        guard let selectionManager: TLSelectionManager = selectionManager as TLSelectionManager? else {
            return NSIndexSet()
        }
        if selectionManager.selectedIndexes != nil {
            return selectionManager.selectedIndexes
        }
        else {
            return NSIndexSet()
        }
    }
    
    func selectedChords() -> [AnyObject] {
        if self.selectedIndexes().count > 0 {
            return tablature!.chordsAtIndexes(self.selectedIndexes())
        }
        else {
            return []
        }
    }
    
    func selectIndexes(indexes: NSIndexSet) {
        selectionManager!.selectIndexes(indexes, byExtendingSelection: false)
    }
    
    func clearSelection() {
        selectionManager!.selectIndexes(NSIndexSet(), byExtendingSelection: false)
    }
    
    // MARK: - Focus handling -
    
    // MARK: Accessors
    
    func focusChord() -> VChord {
        return tablature!.objectInChordsAtIndex(currFocusChordIndex) as! VChord
    }
    
    func focusNote() -> VNote {
        if tabController!.isInSoloMode() {
            return VNote.noteAtFret(VNote.NO_FRET())
        }
        else {
            return self.focusChord()[focusNoteString]!
        }
    }
    
    // MARK: Mutators
    
    func focusNextChord() {
        currFocusChordIndex++
        self.clearSelection()
    }
    
    func focusPrevChord() {
        currFocusChordIndex--
        self.clearSelection()
    }
    
    func focusUpString() {
        focusNoteString--
    }
    
    func focusDownString() {
        focusNoteString++
    }
    
    func reFocusAtPoint(thePoint: NSPoint) {
        var currFocusChordIndex = self.chordIndexAtPoint(thePoint)!
        if currFocusChordIndex >= tablature!.countOfChords() {
            currFocusChordIndex = tablature!.countOfChords() - 1
        }
        focusNoteString = self.stringAtPoint(thePoint)
        self.needsDisplay = true
    }
    
    // MARK: - Input handling -
    
    func handleBoundKey(keyString: String) {
        guard let actionParts = tabController!.keyBindings!.objectForKey(keyString) as? NSArray else {
            return
        }
        let selectorString: String = actionParts.objectAtIndex(0) as! String
        let editSelector: Selector = NSSelectorFromString(selectorString)
        switch actionParts.count {
        case 1:
            tabController!.performSelector(editSelector)
        case 2:
            if selectorString == "addOpenString:" {
                tabController!.addOpenString(actionParts.objectAtIndex(1) as! NSNumber, reverseString: true)
            }
            else {
                tabController!.performSelector(editSelector, withObject: actionParts.objectAtIndex(1))
            }
        case 3:
            if selectorString == "addNoteOnString:onFret:" {
                tabController!.addNoteAtFocus(onString: actionParts.objectAtIndex(1) as! NSNumber, onFret: actionParts.objectAtIndex(2) as! NSNumber, reverseString: true)
            }
            else {
                tabController!.performSelector(editSelector, withObject: actionParts.objectAtIndex(1), withObject: actionParts.objectAtIndex(2))
            }
        default:
            return
        }
    }
    
    public override func keyDown(theEvent: NSEvent) {
        if tabController!.keyBindings!.objectForKey(theEvent.characters!) != nil {
            self.handleBoundKey(theEvent.characters!)
        }
        else {
            self.interpretKeyEvents(NSArray(object: theEvent) as! [NSEvent])
        }
        self.needsDisplay = true
    }
    
    public override func mouseDown(theEvent: NSEvent) {
        lastFocusChordIndex = currFocusChordIndex
        self.mouseDownEvent = theEvent
        selectionManager!.mouseDown(theEvent, userInfo: nil)
        self.reFocusAtPoint(self.convertPoint(theEvent.locationInWindow, fromView: nil))
    }
    
    public override func mouseDragged(theEvent: NSEvent) {
        selectionManager!.mouseDragged(theEvent, userInfo: nil)
        self.reFocusAtPoint(self.convertPoint(theEvent.locationInWindow, fromView: nil))
    }
    
    public override func mouseUp(theEvent: NSEvent) {
        guard let selectionManager = self.selectionManager else {
            return
        }
        selectionManager.mouseUp(theEvent)
        let selection = selectionManager.selectedIndexes
        guard selection != nil else {
            return
        }
        if selection.count == 1 && selection.firstIndex != lastFocusChordIndex {
            self.clearSelection()
        }
        self.reFocusAtPoint(self.convertPoint(theEvent.locationInWindow, fromView: nil))
    }
    
    // MARK: - TLSelectionManager delegate methods -
    
    public func indexUnderPoint(windowPoint: NSPoint, withSelectionManager manager: TLSelectionManager!, userInfo: UnsafeMutablePointer<Void>) -> Int? {
        return self.chordIndexAtPoint(self.convertPoint(windowPoint, fromView: nil))
    }
    
    public override func indexesInBox(windowRect: NSRect, withSelectionManager manager: TLSelectionManager!, userInfo: UnsafeMutablePointer<Void>) -> NSIndexSet? {
        let viewRect: NSRect = self.convertRect(windowRect, fromView: nil)
        let topLeft: NSPoint = viewRect.origin
        let topRight: NSPoint = NSMakePoint(topLeft.x + viewRect.size.width, viewRect.origin.y)
        let bottomLeft: NSPoint = NSMakePoint(viewRect.origin.x, topLeft.y + viewRect.size.height)
        let bottomRight: NSPoint = NSMakePoint(topLeft.x + viewRect.size.width, topLeft.y + viewRect.size.height)
        let mouseDownPoint: NSPoint = self.convertPoint(self.mouseDownEvent!.locationInWindow, fromView: nil)
        var startIndex: Int
        var endIndex: Int
        if NSEqualPoints(mouseDownPoint, topLeft) || NSEqualPoints(mouseDownPoint, bottomRight) {
            startIndex = max(self.chordIndexAtPoint(topLeft)!, 0)
            endIndex = min(self.chordIndexAtPoint(bottomRight)!, tablature!.countOfChords() - 1)
        }
        else {
            startIndex = max(self.chordIndexAtPoint(topRight)!, 0)
            endIndex = min(self.chordIndexAtPoint(bottomLeft)!, tablature!.countOfChords() - 1)
        }
        return NSIndexSet(indexesInRange: NSMakeRange(min(startIndex, endIndex), labs(endIndex - startIndex) + 1))
    }
    
    func shouldResetExistingSelection() -> Bool {
        return true
    }
    
    public override func selectionManagerDidChangeSelection(manager: TLSelectionManager) {
        self.needsDisplay = true
    }
}